src/libcollections/btree/map.rs

```diff
@@ -529,6 +537,174 @@ impl<K: Ord, V> BTreeMap<K, V> {
+                            loop {
+                                match cur_node.ascend() {
+                                    Ok(parent) => {
+                                        let mut parent = parent.into_node();
+                                        if parent.len() < node::CAPACITY {
+                                            // Found a node with space left, push key-value pair and
+                                            // new right subtree.
+                                            let tree_height = parent.height() - 1;
+                                            let mut right_tree = node::Root::new_leaf();
+                                            for _ in 0..tree_height {
+                                                right_tree.push_level();
+                                            }
+
+                                            parent.push(key, value, right_tree);
+                                            cur_node = parent.last_edge().descend();
+                                            break;
+                                        } else {
+                                            // Go up again.
+                                            cur_node = parent.forget_type();
+                                        }
+                                    },
+                                    Err(node) => {
+                                        // We are at the top, create a new root node and subtree.
+                                        let tree_height = node.height();
+                                        let mut new_root = node.into_root_mut().push_level();
+
+                                        let mut right_tree = node::Root::new_leaf();
+                                        for _ in 0..tree_height {
+                                            right_tree.push_level();
+                                        }
+
+                                        // Push key-value pair and new subtree.
+                                        new_root.push(key, value, right_tree);
+                                        cur_node = new_root.last_edge().descend();
+                                        break;
+                                    },
+                                }
+                            }
+
+                            // Go down to the right-most leaf again.
+                            loop {
+                                match cur_node.force() {
+                                    Leaf(leaf) => {
+                                        cur_node = leaf.forget_type();
+                                        break;
+                                    },
+                                    Internal(internal) => cur_node = internal.last_edge().descend(),
+                                }
+                            }
+                        }
+                    },
+                    Internal(_) => unreachable!(),
```

Comment:
This unreachable branch seems a bit ugly to me. It would be better if `cur_node` just explicitly stored a `Leaf` node. The loop for descending to a leaf node would just need to use a separate variable for its internal state.